drivers/usb/udc/udc_smartbond.c

```diff
@@ -0,0 +1,1764 @@
+/*
+ * Copyright (c) 2024 Renesas Electronics Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/*
+ * Driver for the Smartbond USB device controller.
+ */
+
+#include <string.h>
+
+#include "udc_common.h"
+
+#include <DA1469xAB.h>
+#include <da1469x_config.h>
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/dma.h>
+#include <zephyr/drivers/usb/udc.h>
+#include <zephyr/drivers/usb/usb_dc.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/clock_control/smartbond_clock_control.h>
+
+#include <zephyr/logging/log.h>
+#include <zephyr/pm/policy.h>
+
+LOG_MODULE_REGISTER(udc_smartbond, CONFIG_UDC_DRIVER_LOG_LEVEL);
+
+#define DT_DRV_COMPAT renesas_smartbond_usbd
+
+#define UDC_IRQ(inst)      DT_INST_IRQ_BY_IDX(inst, 0, irq)
+#define UDC_IRQ_PRI(inst)  DT_INST_IRQ_BY_IDX(inst, 0, priority)
+#define VBUS_IRQ(inst)     DT_INST_IRQ_BY_IDX(inst, 1, irq)
+#define VBUS_IRQ_PRI(inst) DT_INST_IRQ_BY_IDX(inst, 1, priority)
+
+/*
+ * Minimal transfer size needed to use DMA. For short transfers
+ * it may be simpler to just fill hardware FIFO with data instead
+ * of programming DMA registers.
+ */
+#define DMA_MIN_TRANSFER_SIZE(inst) DT_INST_PROP(inst, dma_min_transfer_size)
+#define FIFO_READ_THRESHOLD(inst)   DT_INST_PROP(inst, fifo_read_threshold)
+
+/* Size of hardware RX and TX FIFO. */
+#define EP0_FIFO_SIZE 8
+#define EP_FIFO_SIZE  64
+
+#define EP_MAX DT_INST_PROP(0, num_bidir_endpoints)
+_Static_assert(EP_MAX > 0 && EP_MAX <= 4, "num-bidir-endpoints out of range");
```

Comment:
Please move it to the bottom, above #define UDC_SMARTBOND_DEVICE_DEFINE(n). Also it does not take instance number in the account, which is probably okay on your platform, but still not correct. Use BUILD_ASSERT. The check is likely not necessary. There other ways to get number of endpoints.